Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,409 @@
+    MultisetDMLT M N := by
+  rcases h with ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  apply MultisetDMLT.DMLT Y {a} X _ M_def N_def
+  · simpa
+  · simp
+
+/- Some useful lemmas. -/
+lemma not_redLT_zero [DecidableEq α] [LT α] (M: Multiset α) : ¬ MultisetRedLT M 0 := by
+  intro h
+  cases h with
+  | RedLT X Y a M nonsense _ =>
+    have contra : a ∈ (0 : Multiset α):= by
+      rw [nonsense]
+      simp_all only [Multiset.mem_add, Multiset.mem_singleton, or_true]
+    contradiction
+
+lemma red_insert [DecidableEq α] [LT α] {a : α} {M N : Multiset α} (h : MultisetRedLT N (a ::ₘ M)) :
+    ∃ M',
+      N = (a ::ₘ M') ∧ (MultisetRedLT M' M) ∨ (N = M + M') ∧ (∀ x : α, x ∈ M' → x < a) := by
+  rcases h with ⟨X, Y, a0, h1, h0, h2⟩
+  by_cases hyp : a = a0
+  · exists Y; right; apply And.intro
+    · rw [h1, add_left_inj]
+      rw [add_comm, Multiset.singleton_add] at h0
+      simp_all
+    · simp_all
+  · exists (Y + (M - {a0}))
+    left
+    constructor
+    · rw [h1]
+      have : X = (M - {a0} + {a}) := by
+        rw [add_comm, Multiset.singleton_add] at *
+        ext b
+        simp [Multiset.ext, Multiset.count_cons] at h0
+        by_cases h : b = a
+        · have := h0 b
+          aesop
+        · have := h0 b
+          simp [Multiset.sub_singleton]
+          aesop
+      subst this
+      rw [add_comm]
+      nth_rewrite 2 [add_comm]
+      rw [Multiset.singleton_add, Multiset.add_cons]
+    · constructor
+      · change Y + (M - {a0}) = (M - {a0}) + Y
+        rw [add_comm]
+      · change M = M - {a0} + {a0}
+        have this0: M = X + {a0} - {a} := by
+          rw [← h0, Multiset.sub_singleton, Multiset.erase_cons_head]
+        have a0M: a0 ∈ M := by
+          rw [this0, Multiset.sub_singleton, Multiset.mem_erase_of_ne]
+          rw [Multiset.mem_add, Multiset.mem_singleton]
+          · apply Or.inr
+            rfl
+          · exact fun h ↦ hyp (Eq.symm h)
+        rw [add_comm]
+        simp_all [Multiset.singleton_add]
+      exact h2
+
+lemma acc_cons [DecidableEq α] [Preorder α] (a : α) (M0 : Multiset α)
+    (_ : ∀ b M , LT.lt b a → AccM M → AccM (b ::ₘ M))
+    (_ : AccM M0)
+    (_ : ∀ M, MultisetRedLT M M0 → AccM (a ::ₘ M)) :
+    AccM (a ::ₘ M0) := by
+  constructor
+  intros N N_lt
+  change AccM N
+  rcases (red_insert N_lt) with ⟨x, H, h0⟩
+  case h.intro.inr h =>
+    rcases h with ⟨H, h0⟩
+    rw [H]
+    clear H
+    induction x using Multiset.induction with
+    | empty =>
+      simpa
+    | cons h =>
+      simp_all only [Multiset.mem_cons, or_true, implies_true, true_implies, forall_eq_or_imp,
+        Multiset.add_cons]
+  case h.intro.inl.intro =>
+    simp_all
+
+lemma acc_cons_of_acc [DecidableEq α] [Preorder α] (a : α)
+    (H : ∀ (b : α), ∀ M, LT.lt b a → AccM M → AccM (b ::ₘ M)) :
+    ∀ M, AccM M → AccM (a ::ₘ M) := by
+  unfold AccM
+  intros M h0
+  induction h0 with
+  | intro x wfH wfh2 =>
+    apply acc_cons
+    · simpa
+    · constructor
+      simpa
+    · simpa
```

Comment:
can these be replaced with something like `constructor <;> simpa`? (don't have an enviroment to test this right now)